pystiche/papers/sanakoyeu_et_al_2018/modules.py

```diff
@@ -0,0 +1,435 @@
+            ),
+            "scale_4": SanakoyeuEtAl2018ConvBlock(
+                in_channels=512,
+                out_channels=512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_5": SanakoyeuEtAl2018ConvBlock(
+                512,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_6": SanakoyeuEtAl2018ConvBlock(
+                1024,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+        }
+    )
+    return modules
+
+
+def sanakoyeu_et_al_2018_prediction_module(
+    in_channels: int, kernel_size: Union[Tuple[int, int], int], padding: str = "same"
+):
+    return sanakoyeu_et_al_2018_conv(
+        in_channels=in_channels,
+        out_channels=1,
+        kernel_size=kernel_size,
+        stride=1,
+        padding=padding,
+    )
+
+
+class SanakoyeuEtAl2018DiscriminatorEncoder(MultiLayerEncoder):
+    def __init__(self, in_channels: int = 3) -> None:
+        super().__init__(
+            self._extract_modules(
+                sanakoyeu_et_al_2018_discriminator_encoder_modules(
+                    in_channels=in_channels
+                )
+            )
+        )
+
+    def _extract_modules(self, wrapped_modules: Dict[str, nn.Sequential]):
+        modules = OrderedDict()
+        block = 0
+        for sequential in wrapped_modules.values():
+            for module in sequential._modules.values():
+                if isinstance(module, nn.Conv2d):
+                    name = f"conv{block}"
+                elif isinstance(module, nn.InstanceNorm2d):
+                    name = f"inst_n{block}"
+                else:  # isinstance(module, nn.LeakyReLU):
+                    name = f"lrelu{block}"
+                    # each LeakyReLU layer marks the end of the current block
+                    block += 1
+
+                modules[name] = module
+        return modules
+
+
+class SanakoyeuEtAl2018Discriminator(object):
+    def __init__(self, in_channels: int = 3) -> None:
+        self.multi_layer_encoder = SanakoyeuEtAl2018DiscriminatorEncoder(
+            in_channels=in_channels
+        )
+        self.prediction_modules = OrderedDict(
+            {
+                "lrelu0": sanakoyeu_et_al_2018_prediction_module(128, 5),
+                "lrelu1": sanakoyeu_et_al_2018_prediction_module(128, 10),
+                "lrelu3": sanakoyeu_et_al_2018_prediction_module(512, 10),
+                "lrelu5": sanakoyeu_et_al_2018_prediction_module(1024, 6),
+                "lrelu6": sanakoyeu_et_al_2018_prediction_module(1024, 3),
+            }
+        )
```

Comment:
This is dangerous: you create an ordered structure (`OrderedDict`) from an unordered structure (`dict`). That means that while `self.prediction_modules` has a deterministic ordering, there is no guarantee that is is the ordering you wrote down. While that should be no problem here, I suggest you correct this anyway. You could do it like this

```python
OrderedDict(
    (
        ("lrelu0", sanakoyeu_et_al_2018_prediction_module(128, 5)),
        ("lrelu1", sanakoyeu_et_al_2018_prediction_module(128, 10)),
        ...,
    )
)
```